src/main/java/prefs/ConfigFileHandler.java

```diff
@@ -12,67 +12,131 @@
+     * @param configDirectory The directory that the config file is held in
+     * @param configFileName The file name of the config file
+     * @param config The config file object to deserialise into.
+     */
+    public ConfigFileHandler(String configDirectory, String configFileName, Config config) {
+        this.configFileName = configFileName;
         this.configDirectory = configDirectory;
+        this.config = config;
 
         setupGson();
         ensureDirectoryExists();
     }
 
-    private void ensureDirectoryExists() {
+    /**
+     * Ensures that the directory specified in the attribute exists.
+     * Attempts to create the directory if it does not exist.
+     * @return True if the directory exists or was created.
+     *         False if by the end of the method, the directory still doesn't exist
+     */
+    private boolean ensureDirectoryExists() {
         File directory = new File(configDirectory);
-        boolean directoryNonExistent = !directory.exists() || !directory.isDirectory();
-        boolean exists = true;
+        boolean exists = directory.exists();
+        boolean isDir = directory.isDirectory();
+        boolean directoryNonExistent = !exists || !isDir;
         if (directoryNonExistent) {
             exists = directory.mkdirs();
         }
         if (!exists) {
             logger.warn("Could not create config file directory");
         }
+        return exists;
     }
 
-    private File getGlobalConfigFile() {
-        return new File(configDirectory, globalConfigFileName).getAbsoluteFile();
+    /**
+     * Returns the File representation at the config file path
+     */
+    private File getConfigFile() {
+        return new File(configDirectory, configFileName).getAbsoluteFile();
     }
 
-    public void saveGlobalConfig(GlobalConfig config) {
-        File configFile = getGlobalConfigFile();
+    /**
+     * Serialises the config attribute object to a config file.
+     */
+    public void saveConfig() {
+        File configFile = getConfigFile();
         try {
             Writer writer = new OutputStreamWriter(new FileOutputStream(configFile) , CHARSET);
-            gson.toJson(config, GlobalConfig.class, writer);
+            if (config == null) {
+                logger.warn("Null config object, simply creating a file with an empty json.");
+                writer.write("{}");
+            } else {
+                gson.toJson(config, config.getClass(), writer);
+            }
             writer.close();
         } catch (IOException e) {
             HTLog.error(logger, e);
         }
     }
 
-    public GlobalConfig initGlobalConfig() {
-        GlobalConfig config = new GlobalConfig();
-        saveGlobalConfig(config);
-        return config;
-    }
+    /**
+     * Initialises the config of the type that was passed in during the constructor.
+     * @return The initialised config object, different from the original config object.
+     */
+    public Config initConfig() {
```

Comment:
In my opinion, this creates more trouble for us *and* I don't think initialising the config is the job of config file handler. Initialising should be the job of whoever is owning the config, which in this case is `Preference`.

So, instead of having the complication of instancing a class, why don't we initialise the config in `Preference` and pass whatever config was created into config file handler constructor. This way we don't need `initConfig` which obviously is not config file handler job, although we still need `loadConfigFromFile`.

What do you think?